regression/polus-bleed-through-estimation-plugin/src/tile_selectors.py

```diff
@@ -0,0 +1,179 @@
+                score, image_min, image_max = future.result()
+                self.__scores.append(score)
+                self.__image_mins.append(image_min)
+                self.__image_maxs.append(image_max)
+
+        self.__selected_tiles = self._select_best_tiles()
+
+    @property
+    def selected_tiles(self) -> utils.TileIndices:
+        return self.__selected_tiles
+
+    @property
+    def image_mins(self) -> list[int]:
+        return self.__image_mins
+
+    @property
+    def image_maxs(self) -> list[int]:
+        return self.__image_maxs
+
+    @abc.abstractmethod
+    def _score_tile(self, tile: numpy.ndarray) -> float:
+        raise NotImplementedError(f'Any subclass of Criterion must implement the \'_score_tile\' method.')
+
+    def _score_tiles_thread(self, file_path: Path) -> tuple[utils.ScoresDict, int, int]:
+        """ This method runs in a single thread and scores all tiles for a single file.
+
+        Args:
+            file_path: Path to image for which the tiles need to be scored.
+
+        Returns:
+            A Dictionary of tile-scores. See `utils/types.py`
+        """
+        with BioReader(file_path, max_workers=utils.NUM_THREADS) as reader:
+
+            scores_dict: utils.ScoresDict = dict()
+            logger.info(f'Ranking tiles in {file_path.name}...')
+            num_tiles = utils.count_tiles(reader)
+            image_min = numpy.iinfo(reader.dtype).max
+            image_max = -numpy.iinfo(reader.dtype).min
+
+            for i, (_, _, y_min, y_max, x_min, x_max) in enumerate(utils.tile_indices(reader)):
+                if i % 10 == 0:
+                    logger.info(f'Ranking tiles in {file_path.name}. Progress {100 * i / num_tiles:6.2f} %')
+
+                tile = numpy.squeeze(reader[y_min:y_max, x_min:x_max, 0, 0, 0])
+
+                # TODO: Actually handle 3d images properly with 3d tile-chunks.
+                key = (0, 1, y_min, y_max, x_min, x_max)
+                if key in scores_dict:
+                    scores_dict[key] = (max if self.__is_high_better else min)(scores_dict[key], self._score_tile(tile))
+                else:
+                    scores_dict[key] = self._score_tile(tile)
+
+                image_min = numpy.min(tile[tile > 0], initial=image_min)
+                image_max = numpy.max(tile, initial=image_max)
+
+        return scores_dict, image_min, image_max
+
+    def _select_best_tiles(self) -> utils.TileIndices:
+        """ Sort the tiles based on their scores and select the best few from each channel
+
+        Returns:
+            List of indices of the best tiles. See `utils.types.py`
+        """
+        return list(set(
+            coordinates for scores_dict in self.__scores
+            for coordinates, _ in list(sorted(
+                scores_dict.items(),
+                key=itemgetter(1),
+                reverse=self.__is_high_better,
+            ))[:self.__num_tiles_per_channel]
+        ))
+
+
+class HighEntropy(Selector):
```

Comment:
For the selector class names, technically you have decoupled the criterion (min vs max) from the measure (entropy, mean intensity, etc...) so I think names such as `Entropy` `MeanIntensity` etc.. should be preferred.